# import libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


# Load data

In [10]:
df=pd.read_csv(r"D:\FCDS\internship\CodeAlpha\sales prediction\Advertising.csv")


# EDA

In [13]:
df.head()

Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [15]:
df.describe()

Unnamed: 0          TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000  200.000000
mean   100.500000  147.042500   23.264000   30.554000   14.022500
std     57.879185   85.854236   14.846809   21.778621    5.217457
min      1.000000    0.700000    0.000000    0.300000    1.600000
25%     50.750000   74.375000    9.975000   12.750000   10.375000
50%    100.500000  149.750000   22.900000   25.750000   12.900000
75%    150.250000  218.825000   36.525000   45.100000   17.400000
max    200.000000  296.400000   49.600000  114.000000   27.000000

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  200 non-null    int64  
 1   TV          200 non-null    float64
 2   Radio       200 non-null    float64
 3   Newspaper   200 non-null    float64
 4   Sales       200 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 7.9 KB


In [19]:
print(df.isnull().sum())

Unnamed: 0    0
TV            0
Radio         0
Newspaper     0
Sales         0
dtype: int64


In [25]:
print(df.duplicated().sum())

0


In [31]:
def detect_outliers_iqr(dataframe, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = dataframe[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return outliers


In [33]:
# Detect outliers in each numerical column
outliers_tv = detect_outliers_iqr(df, 'TV')
outliers_radio = detect_outliers_iqr(df, 'Radio')
outliers_newspaper = detect_outliers_iqr(df, 'Newspaper')
outliers_sales = detect_outliers_iqr(df, 'Sales')

In [39]:
# Combine all detected outliers into a single DataFrame without duplicates
outliers_all = pd.concat([
    outliers_tv, outliers_radio, outliers_newspaper, outliers_sales
]).drop_duplicates()

print(outliers_combined)

     Unnamed: 0     TV  Radio  Newspaper  Sales
16           17   67.8   36.6      114.0   12.5
101         102  296.4   36.3      100.9   23.8


In [41]:
df_cleaned = df.drop(outliers_all.index)

In [45]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 198 entries, 0 to 199
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  198 non-null    int64  
 1   TV          198 non-null    float64
 2   Radio       198 non-null    float64
 3   Newspaper   198 non-null    float64
 4   Sales       198 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 9.3 KB


# Preprocessing

In [48]:
X = df[['TV', 'Radio', 'Newspaper']]
y = df['Sales']

In [50]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Model Training

In [53]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [55]:
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [57]:
print("\nModel Coefficients:", model.coef_)
print("Model Intercept:", model.intercept_)
print("Mean Squared Error:", mse)
print("R² Score:", r2)


Model Coefficients: [0.04472952 0.18919505 0.00276111]
Model Intercept: 2.979067338122629
Mean Squared Error: 3.1740973539761046
R² Score: 0.899438024100912



## The model performs very well with a high R² score and low error rate.

## Radio advertising contributes most to increasing sales.

## Newspaper ads have negligible impact and may be reconsidered in future campaigns.

The intercept of 2.979 implies baseline sales of ~3 units with no advertising.

